### PR TITLE
Improve dim label map

### PR DESCRIPTION
**Pull request by SimonHeybrock**
As discussed:
- Use `std::shared_mutex`, since mostly we are just reading.
- Remove neutron spefici labels. Will define these in `scippneutron` (only relevant for use from C++).

### CODE REVIEW

File: units/dim.cpp [hunk 1]
```diff
@@ -3,49 +3,36 @@
...
 /// @author Simon Heybrock
 #include <limits>
+#include <mutex>
 
 #include "scipp/common/index.h"
 #include "scipp/units/dim.h"
 
 namespace scipp::units {
 
 std::unordered_map<std::string, Dim::Id> Dim::builtin_ids{
```
Comment (jl-wynen) [inline units/dim.cpp:13]: Does it make sense to keep the two separate maps? How expensive is the lookup in `builtin_ids` compared to acquiring the shared lock?

File: units/dim.cpp [hunk 2]
```diff
@@ -55,6 +42,8 @@ Dim::Dim(const std::string &label) {
     throw std::runtime_error(
         "Exceeded maximum number of different dimension labels.");
   m_id = static_cast<Id>(id);
+  read_lock.unlock();
+  const std::unique_lock write_lock(mutex);
   custom_ids[label] = m_id;
```
Comment (jl-wynen) [inline units/dim.cpp:47]: This does not work when two threads want to insert two different ids at the same time.
Suppose thread 1 and 2 are both in the reading block with 2 different labels. Neither finds its label in `custom_ids` so both compute `id = size(custom_ids) + 1000` with the __same__ size.

You have to consider everything after the `custom_ids` lookup as part of the write block.

**Review by jl-wynen — CHANGES_REQUESTED** [review]

File: units/dim.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply units/dim.cpp:13]: I don't know. I suppose we can figure that out separately?

File: units/dim.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply units/dim.cpp:47]: Right, I need look write-lock earlier.

Comment (SimonHeybrock) [reply units/dim.cpp]: Good catch!

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**